scheduler/src/cook/mesos/util.clj

```diff
@@ -391,30 +453,37 @@
 ;; running and waiting jobs, so it's implemented that way to keep things simple.
 (defn get-active-jobs-by-user-and-state
   "Returns all jobs for a particular user in the specified state
-   and timeframe, without a custom executor.
-   Note that this query is not performant for completed jobs, use
-   get-completed-jobs-by-user instead."
-  [db user start end state name-filter-fn include-custom-executor?]
+   and timeframe, without a custom executor. This query works for state
+   waiting and running only."
+  [db user start end limit state name-filter-fn include-custom-executor?]
   (let [state-keyword (case state
                         "running" :job.state/running
                         "waiting" :job.state/waiting)
-        jobs
-        (timers/time!
-          (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-duration")])
-          (->> (q '[:find [?j ...]
-                    :in $ ?user ?state ?start ?end
-                    :where
-                    [?j :job/state ?state]
-                    [?j :job/user ?user]
-                    [?j :job/submit-time ?t]
-                    [(<= ?start ?t)]
-                    [(< ?t ?end)]]
-                  db user state-keyword start end)
-               (map (partial d/entity db))))]
-    (cond->> jobs
-             (not include-custom-executor?) (filter #(false? (:job/custom-executor %)))
-             name-filter-fn (filter #(name-filter-fn (:job/name %)))
-             (and include-custom-executor? (= :job.state/waiting state-keyword)) (remove uncommitted?))))
+        ;; This is used to compute the histogram bucket to put in. take the time range in hours, with a one minute leeway. E.g., this counts
+        ;; a requested time of 24 hours + 3 seconds into the 24-hour bucket.
+        millis-diff (max 0 (- (.getTime end) (.getTime start) (-> 1 t/minutes t/in-millis)))
+        hours-diff (/ millis-diff (-> 1 t/hours t/in-millis))
+        ; A histogram bucket, so that we can see the query time distribution as a function of the range of time being searched over.
+        bucket (cond
+                 (<= hours-diff 6) 6
+                 (<= hours-diff 12) 12
+                 (<= hours-diff 24) 24
+                 (<= hours-diff 36) 36
+                 (<= hours-diff 48) 48
+                 (<= hours-diff 72) 72
+                 :else 999)]
+    (timers/time!
+      (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-duration")])
+      (timers/time!
+        (timers/timer ["cook-mesos" "scheduler" (str "get-" (name state) "-jobs-by-user-by-state-" bucket "-duration")])
```

Comment:
Why are we doing the bucket timer for active but not for completed?

Reply:
We can as part of https://github.com/twosigma/Cook/issues/780. Right now, no point, completed jobs metric is junk.

Reply:
Doesn't it make sense to introduce the bucket timers at the same time for active and completed though?

Reply:
We need to inject doall's in completed in the right spots so that we get good answers. I agree that we want to do the work, but don't want to grow the scope of this JIRA.

Reply:
I would vote to make this consistent as part of this PR, in the sense that we either have the bucket timers for both or for neither.

Reply:
I also think this could be broken out into a helper function if we were doing both at the same time. I'd probably vote to just leave it out for now.

Reply:
Code ripped out.